e2e/playground.py

```diff
@@ -0,0 +1,2135 @@
+        self.date_end.expect_max_date(value, timeout=timeout)
+
+    # format: str = "yyyy-mm-dd",
+    def expect_format(
+        self,
+        value: AttrValue,
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        # These values should be the same, so checking both independently seems fair
+        self.date_start.expect_format(value, timeout=timeout)
+        self.date_end.expect_format(value, timeout=timeout)
+
+    # startview: str = "month",
+    def expect_startview(
+        self,
+        value: AttrValue,
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        # These values should be the same, so checking both independently seems fair
+        self.date_start.expect_startview(value, timeout=timeout)
+        self.date_end.expect_startview(value, timeout=timeout)
+
+    # weekstart: int = 0,
+    def expect_weekstart(
+        self,
+        value: typing.Union[int, AttrValue],
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        # These values should be the same, so checking both independently seems fair
+        self.date_start.expect_weekstart(value, timeout=timeout)
+        self.date_end.expect_weekstart(value, timeout=timeout)
+
+    # language: str = "en",
+    def expect_language(
+        self,
+        value: AttrValue,
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        # These values should be the same, so checking both independently seems fair
+        self.date_start.expect_language(value, timeout=timeout)
+        self.date_end.expect_language(value, timeout=timeout)
+
+    # separator: str = " to ",
+    def expect_separator(
+        self,
+        value: TextValue,
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        if value is None:
+            value = ""
+        playwright_expect(self.loc_separator).to_have_text(value, timeout=timeout)
+
+    # width: Optional[str] = None,
+
+    # autoclose: bool = True,
+    def expect_autoclose(
+        self,
+        value: bool,
+        *,
+        timeout: Timeout = None,
+    ) -> None:
+        # These values should be the same, so checking both independently seems fair
+        self.date_start.expect_autoclose(value, timeout=timeout)
+        self.date_end.expect_autoclose(value, timeout=timeout)
+
+
+######################################################
+# # Outputs
+######################################################
+
+
+class _OutputBaseP(Protocol):
+    id: str
+    loc: Locator
+
+
+class _OutputBase:
+    id: str
+    loc: Locator
+
+    def __init__(
+        self,
+        page: Page,
+        *,
+        id: str,
+        loc: InitLocator,
+    ) -> None:
+        self.page = page
+        self.id = id
+
+        if isinstance(loc, str):
+            loc = page.locator(loc)
+        self.loc = loc
+
+    @property
+    # TODO; Return type
+    def expect(self):
+        return playwright_expect(self.loc)
+
+
+class _OutputTextValue(_OutputBase):
```

Comment:
We should be able to run assertions for:

0. It has not received its initial value
1. It's showing a valid output result
    1. Optionally, pattern or string match on the result
2. It's showing an error
    1. Optionally, pattern or string match on the result
3. It's in progress

0, 2, and 3 should be added to `OutputBase` because they don't vary.

(This is another candidate for composing assertions)

Reply:
Made issue: https://github.com/rstudio/py-shiny/issues/409

Resolving